src/tss2-tcti/tcti-libtpms.c

```diff
@@ -0,0 +1,317 @@
+{
+    if (tcti_ctx == NULL)
+        return NULL;
+
+    return (TSS2_TCTI_LIBTPMS_CONTEXT*)tcti_ctx;
+}
+/*
+ * This function down-casts the libtpms TCTI context to the common context
+ * defined in the tcti-common module.
+ */
+TSS2_TCTI_COMMON_CONTEXT*
+tcti_libtpms_down_cast(TSS2_TCTI_LIBTPMS_CONTEXT *tcti_libtpms)
+{
+    if (tcti_libtpms == NULL) {
+        return NULL;
+    }
+    return &tcti_libtpms->common;
+}
+
+TSS2_RC
+tcti_libtpms_transmit(
+    TSS2_TCTI_CONTEXT *tcti_ctx,
+    size_t size,
+    const uint8_t *cmd_buf)
+{
+    TSS2_TCTI_LIBTPMS_CONTEXT *tcti_libtpms = tcti_libtpms_context_cast(tcti_ctx);
+    TSS2_TCTI_COMMON_CONTEXT *tcti_common = tcti_libtpms_down_cast(tcti_libtpms);
+    tpm_header_t header;
+    TSS2_RC rc;
+    TPM_RESULT ret;
+
+    rc = tcti_common_transmit_checks(tcti_common, cmd_buf, TCTI_LIBTPMS_MAGIC);
+    if (rc != TSS2_RC_SUCCESS) {
+        return rc;
+    }
+    rc = header_unmarshal(cmd_buf, &header);
+    if (rc != TSS2_RC_SUCCESS) {
+        return rc;
+    }
+    if (header.size != size) {
+        LOG_ERROR ("Buffer size parameter: %zu, and TPM2 command header size "
+                   "field: %" PRIu32 " disagree.", size, header.size);
+        return TSS2_TCTI_RC_BAD_VALUE;
+    }
+
+    LOG_DEBUG("Sending command with TPM_CC 0x%" PRIx32 " and size %" PRIu32,
+               header.code, header.size);
+
+    ret = tcti_libtpms->TPMLIB_Process(&tcti_libtpms->response_buffer,
+                                       (uint32_t *) &tcti_libtpms->response_len,
+                                       (uint32_t *) &tcti_libtpms->response_buffer_len,
+                                       (uint8_t *) cmd_buf,
+                                       size);
+    if (ret != TPM_SUCCESS) {
+        LOG_ERROR("libtpms function TPMLIB_Process() failed with return code 0x%" PRIx32, ret);
+        return TSS2_TCTI_RC_IO_ERROR;
+    }
+
+    tcti_common->state = TCTI_STATE_RECEIVE;
+
+    return TSS2_RC_SUCCESS;
+}
+
+TSS2_RC
+tcti_libtpms_cancel(
+    TSS2_TCTI_CONTEXT *tctiContext)
+{
+    (void) (tctiContext);
+    return TSS2_TCTI_RC_NOT_IMPLEMENTED;
+}
+
+TSS2_RC
+tcti_libtpms_set_locality(
+    TSS2_TCTI_CONTEXT *tctiContext,
+    uint8_t locality)
+{
+    (void) (tctiContext);
+    (void) (locality);
+    return TSS2_TCTI_RC_NOT_IMPLEMENTED;
+}
+
+TSS2_RC
+tcti_libtpms_get_poll_handles(
+    TSS2_TCTI_CONTEXT *tctiContext,
+    TSS2_TCTI_POLL_HANDLE *handles,
+    size_t *num_handles)
+{
+    (void)(tctiContext);
+    (void)(handles);
+    (void)(num_handles);
+    return TSS2_TCTI_RC_NOT_IMPLEMENTED;
+}
+
+void
+tcti_libtpms_finalize(
+    TSS2_TCTI_CONTEXT *tctiContext)
+{
+    TSS2_TCTI_LIBTPMS_CONTEXT *tcti_libtpms = tcti_libtpms_context_cast(tctiContext);
+    TSS2_TCTI_COMMON_CONTEXT *tcti_common = tcti_libtpms_down_cast(tcti_libtpms);
+
+    if (tcti_libtpms == NULL) {
+        return;
+    }
+
+    if (tcti_common->state == TCTI_STATE_RECEIVE) {
```

Comment:
Maybe that should be if (!tcti_libtpms->response_buffer)

Reply:
Agreed, that would be more elegant.